cpu/stm32f2/periph/spi.c

```diff
@@ -397,28 +429,142 @@ void spi_poweroff(spi_t dev)
+            while (SPI_1_DEV->SR & SPI_SR_BSY) {} 
             SPI_1_CLKDIS();
             break;
 #endif
 #if SPI_2_EN
         case SPI_2:
-            while (SPI_2_DEV->SR & SPI_SR_BSY);
+            while (SPI_2_DEV->SR & SPI_SR_BSY) {}
             SPI_2_CLKDIS();
             break;
 #endif
     }
 }
 
-static inline void irq_handler_transfer(SPI_TypeDef *spi, spi_t dev)
+int spi_transfer_bytes(spi_t dev, char *out, char *in, unsigned int length)
 {
+    SPI_TypeDef *spi_dev;
+    uint32_t tx_stream;
+    uint32_t rx_stream;
+    uint32_t tx_channel;
+    uint32_t rx_channel;
+    uint32_t tx_conf;
+    uint32_t rx_conf;
+    uint32_t periph_addr_reg;
+    char temp_location;
 
+    switch (dev) {
+#if SPI_0_EN
+        case SPI_0:
+            spi_dev = SPI_0_DEV;
+
+            tx_stream = SPI_0_DMA_TX_STREAM;
+            rx_stream = SPI_0_DMA_RX_STREAM;
+            tx_channel = SPI_0_DMA_TX_CHANNEL;
+            rx_channel = SPI_0_DMA_RX_CHANNEL;
+            break;
+#endif  /* SPI_0_EN */
+#if SPI_1_EN
+        case SPI_1:
+            spi_dev = SPI_1_DEV;
+
+            tx_stream = SPI_1_DMA_TX_STREAM;
+            rx_stream = SPI_1_DMA_RX_STREAM;
+            tx_channel = SPI_1_DMA_TX_CHANNEL;
+            rx_channel = SPI_1_DMA_RX_CHANNEL;
+            break;
+#endif  /* SPI_1_EN */
+#if SPI_2_EN
+        case SPI_2:
+            spi_dev = SPI_2_DEV;
+
+            tx_stream = SPI_2_DMA_TX_STREAM;
+            rx_stream = SPI_2_DMA_RX_STREAM;
+            tx_channel = SPI_2_DMA_TX_CHANNEL;
+            rx_channel = SPI_2_DMA_RX_CHANNEL;
+            break;
+#endif  /* SPI_2_EN */
+        default:
+            return -1;
+    }
+
+    periph_addr_reg = (uint32_t)&(spi_dev->DR);
+
+    /* DMA configuration */
+    if (in == NULL) {
+        /* channel - medium priority - peripheral to memory - complete interrupt enable */
+        rx_conf = ((rx_channel << 25) | DMA_SxCR_PL_0 | DMA_SxCR_TCIE);
+        in = &temp_location;
+    }
+    else {
+        /* channel - medium priority - peripheral to memory - memory increment - complete interrupt enable */
+        rx_conf = ((rx_channel << 25) | DMA_SxCR_PL_0 | DMA_SxCR_MINC | DMA_SxCR_TCIE);
+    }
+
+    if (out == NULL) {
+        /* channel - medium priority - memory to peripheral - complete interrupt enable*/
+        tx_conf = ((tx_channel << 25) | DMA_SxCR_PL_0 | DMA_SxCR_DIR_0 | DMA_SxCR_TCIE);
+        out = &temp_location;
+    }
+    else {
+        /* channel - medium priority - memory to peripheral - memory increment - complete interrupt enable */
+        tx_conf = ((tx_channel << 25) | DMA_SxCR_PL_0 | DMA_SxCR_DIR_0 | DMA_SxCR_MINC | DMA_SxCR_TCIE);
+    }
+
+    /* acquire dma spi rx and tx streams */
+    dma_conf_acquire(tx_stream);
+    dma_conf_acquire(rx_stream);
+
+    /* streams configuration */
+    dma_stream_config(tx_stream, periph_addr_reg, tx_conf, out, (uint16_t)length);
+    dma_stream_config(rx_stream, periph_addr_reg, rx_conf, in, (uint16_t)length);
```

Comment:
Try and use modified function prototype. I think GCC's `LTO` is screwing this up :)